ghcup.cabal

```diff
@@ -111,6 +111,8 @@ library
   build-depends:
     , aeson                 >=1.4
     , async                 >=0.8        && <2.3
+    , autodocodec           ^>=0.2.0.3
```

Comment:
I'm **very** mindful about the dependency footprint of GHCup. It doesn't seem we need or want this dependency for reasons other than generating the schema? Can we do with CPP and a flag?

Also... we should have a binary that generates the actual schema. Preferably in ghcup-metadata repo: https://github.com/haskell/ghcup-metadata/tree/develop/ghcup-gen

Reply:
Autodocodec also provides the `FromJSON` and `ToJSON` instances, although obviously you can get those with Aeson alone. So in that sense it is just for the schema. 

You could hide the dependency behind a flag, but at that point I think you'd be better off simply adding a `schema.json` file to the repo somewhere. The benefit of Autodocodec is making sure that the instances all agree with each other. If it's only used conditionally, you lose that benefit. 

I also try to avoid heavy dependencies. I think the only thing that Autodocodec adds beyond Aeson's dependencies is [the `validity` package](https://hackage.haskell.org/package/validity-0.12.0.1). 

`autodocodec-schema`, on the other hand, could certainly be hidden away behind a flag. Its only purpose is to turn the `Codec` into an actual schema. Perhaps it could be part of the test suite, as a golden test?

Reply:
I'm not a big believer in autogenerating schemas like people do with servant and swagger.

A schema is a user facing document that needs to be designed. Then we want to make sure our metadata files always adhere to it.

So:

- we have a metadata file
- ghcup can parse it (already checked)
- we have a schema
- our metadata files match the schema at all times (needs to be added to CI)

I'd also say we want to turn the json schema into yaml, so we can add comments etc.

At that stage, we won't really need any generation mechanism, I guess.

Reply:
In that case, I'll close this PR. I posted the generated schema in the description — hopefully it can be used as a starting point. 

I've used [AJV](https://ajv.js.org) for validating schemas in CI and it's worked well for me.